### PR TITLE
CDRIVER-3917 route all test client and client_pool creation through test framework

### PR DESCRIPTION
**Pull request by samantharitter**
This PR renames some of the test framework methods around client and client_pool creation so they match the signatures of the `mongoc_client_new` and `mongoc_client_pool_new` methods.  It also creates new methods in the test framework for client and pool creation, so that all clients and pools in our tests are created through the framework.

The framework stub methods right now have simple pass through to `mongoc_client_new` and `mongoc_client_pool_new`, so these changes should have no behavior changes in our tests.  The next piece of work for CDRIVER-3917 will be to change these stub methods to add in versioned API information to client and pool creation.

### CODE REVIEW

File: src/libmongoc/tests/test-libmongoc.h [hunk 1]
```diff
@@ -98,10 +98,18 @@ void
...
+test_framework_client_new (const char *uri_str);
 mongoc_client_t *
-test_framework_client_new (void);
+test_framework_client_new_from_uri (const mongoc_uri_t *uri);
+
+mongoc_client_pool_t *
+test_framework_new_default_client_pool (void);
 mongoc_client_pool_t *
-test_framework_client_pool_new (void);
+test_framework_client_pool_new (const mongoc_uri_t *uri);
```
Comment (alcaeus) [inline src/libmongoc/tests/test-libmongoc.h:112]: I dislike that this does not match the signature of `test_framework_client_new` but rather `test_framework_client_new_from_uri`.

My suggestion would be:
```suggestion
test_framework_client_pool_new (const char *uri_str);
mongoc_client_t *
test_framework_client_pool_new_from_uri (const mongoc_uri_t *uri);
```

**Review by alcaeus — COMMENTED** [review]
I didn't bother to look at the result of the refactoring, so I'll let evergreen decide on that.

File: src/libmongoc/tests/test-libmongoc.h (hunk 1 above)
Comment (samantharitter) [reply src/libmongoc/tests/test-libmongoc.h:112]: The reason I changed this to take a uri was to match the signature of `mongoc_client_pool_new`, which takes a uri:

```
MONGOC_EXPORT (mongoc_client_pool_t *)
mongoc_client_pool_new (const mongoc_uri_t *uri);
```

I find it confusing that the test framework doesn't match the real class' methods.  It's also a bit weird that `mongoc_client_pool_t` and `mongoc_client_t` have methods whose signatures don't agree, but that's a discussion for another day.

Do you think it's more important that `test_framework_client_pool_*` agrees with `mongoc_client_pool_*`, or that it agrees with `test_framework_client_*` ? Unfortunately we can't have both :(

Comment (alcaeus) [reply src/libmongoc/tests/test-libmongoc.h:112]: I think it's more natural to have the `test_framework_client_*` methods have consistent signatures and deal with any API differences internally. This at least fixes the wrong signature issue in tests, leaving us to figure out the actual API on a different day.

**Review by alcaeus — APPROVED** [review]

**Comment by kevinAlbs** [comment]
Being done in #787.

**Pull request closed without merging**